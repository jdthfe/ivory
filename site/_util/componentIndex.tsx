export default [
    { name: 'CheckUA', type: 'Function', 'zh-CN': '校验UA' },
    { name: 'DeepObjectMerge', type: 'Function', 'zh-CN': '数据扁平化' },
    { name: 'FormatePrice', type: 'Function', 'zh-CN': '格式化价钱' },
    { name: 'GetQuery', type: 'Function', 'zh-CN': '解析URL' },
    { name: 'Throttle', type: 'Function', 'zh-CN': '待输入_中文名' },
];
