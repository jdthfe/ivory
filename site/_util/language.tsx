// en-US must be the first item
export interface LanguageItem {
    code: Language;
    content: string;
}
export const languages: LanguageItem[] = [
    { code: 'en-US', content: 'English' },
    { code: 'zh-CN', content: '中文' },
];

/** 判断语言 */
export function getLanguage(): Language {
    const code = String(localStorage.getItem('language')) as Language;
    if (languages.map((item) => item.code).includes(code)) {
        return code;
    } else {
        return 'en-US';
    }
}
