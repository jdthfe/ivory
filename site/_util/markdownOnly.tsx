export default [
    { name: 'Introduce', type: 'markdownOnly', 'zh-CN': '介绍' },
    { name: 'Customize Theme', type: 'markdownOnly', 'zh-CN': '主题' },
    { name: '_util', type: 'markdownOnly', 'zh-CN': '工具' },
];
