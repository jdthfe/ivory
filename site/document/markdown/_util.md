Common variables in components

```tsx
export const prefix = 'eui';
export const Variable = {
    defaultDuration: 2000,
    transitionTime: 160,
    transitionFade: `${prefix}-fade`,
    transitionZoom: `${prefix}-zoom`,
};

export const removeNode = (el: HTMLDivElement) => {
    if (el.remove) {
        el.remove();
    } else if (el.parentNode) {
        el.parentNode.removeChild(el);
    }
};

export default {
    prefix,
    Variable,
    removeNode,
};
```

Usage

```tsx
import { _util } from '@jdthfe/eui';
// import _util from 'eui/compiled/lib/_util/'; // If you use babel-plugin-import

const { Variable } = _util;
console.log(Variable.defaultDuration); // 2000
```
