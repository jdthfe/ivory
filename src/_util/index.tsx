export const prefix = 'eui';
export const Variable = {
    defaultDuration: 2000,
    transitionTime: 160,
    transitionFade: `${prefix}-fade`,
    transitionZoom: `${prefix}-zoom`,
};

export const removeNode = (el: HTMLDivElement) => {
    if (el.remove) {
        el.remove();
    } else if (el.parentNode) {
        el.parentNode.removeChild(el);
    }
};

export default {
    prefix,
    Variable,
    removeNode,
};
