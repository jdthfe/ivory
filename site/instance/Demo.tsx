import React from 'react';
import { getInstance } from '../_util';

interface DemoProps {
    item: ComponentIndex;
}
const Demo = (props: DemoProps) => {
    const { item } = props;
    const Demo = getInstance(item) || null;
    return <main className="instance-demo">{Demo ? <Demo /> : null}</main>;
};

export default Demo;
