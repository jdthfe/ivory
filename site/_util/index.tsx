import componentIndexNoType from './componentIndex';
import markdownOnlyNoType from './markdownOnly';
import { languages } from './language';

export interface Docs {
    readme: string;
    'readme.zh-CN'?: string;
    demoSource?: string;
    [key: string]: string | undefined;
}

export const componentIndex: ComponentIndex[] = componentIndexNoType;
export const markdownOnly: ComponentIndex[] = markdownOnlyNoType;

export const sortComponentIndex = (componentIndex: ComponentIndex[]) => {
    const newStructure: { [propName: string]: ComponentIndex[] } = {};
    componentIndex.map((item) => {
        if (newStructure[item.type]) {
            newStructure[item.type].push(item);
        } else {
            newStructure[item.type] = [item];
        }
    });
    return newStructure;
};
export const getInstance = (item: ComponentIndex) => {
    let instance: React.FC;
    try {
        if (item.type === 'markdownOnly') {
            throw 'markdownOnly have not Instance';
        } else {
            instance = require(`../../src/${item.name}/demo`).default;
        }
    } catch (e) {
        throw e;
    }
    return instance;
};
export const getDocs = (item: ComponentIndex) => {
    const docs: Docs = { readme: '' };
    try {
        if (item.type === 'markdownOnly') {
            languages.map((lang) => {
                const readmeLang = `readme${
                        lang.code === 'en-US' ? '' : '.' + lang.code
                    }`,
                    mdLang = `${item.name}${
                        lang.code === 'en-US' ? '' : '.' + lang.code
                    }`;
                docs[readmeLang] =
                    require(`!!raw-loader!../document/markdown/${mdLang}.md`).default;
            });
        } else {
            docs.demoSource =
                require(`!!raw-loader!../../src/${item.name}/demo`).default;
            languages.map((lang) => {
                const readmeLang = `readme${
                    lang.code === 'en-US' ? '' : '.' + lang.code
                }`;
                docs[readmeLang] =
                    require(`!!raw-loader!../../src/${item.name}/demo/${readmeLang}.md`).default;
            });
        }
    } catch (e) {
        throw e;
    }
    return docs;
};
