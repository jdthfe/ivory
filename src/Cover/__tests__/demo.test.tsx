import webDemoTest from '@tests/shared/demoTest';
webDemoTest('Cover');

import React from 'react';
import Demo from '../demo';
import { Variable } from '../../_util/';
const { transitionFade } = Variable;
import { render, fireEvent, waitForElement } from 'react-testing-library';

// https://codesandbox.io/s/rqj0lymyn?from-embed
// https://reactjs.org/docs/hooks-faq.html#how-to-test-components-that-use-hooks
// https://codesandbox.io/s/xrw09qlx4o
// https://github.com/facebook/react/issues/15379

describe('CoverDemo', () => {
    it('Cover change visible', async () => {
        const { baseElement, getAllByRole } = render(<Demo />);

        // https://testing-library.com/docs/dom-testing-library/cheatsheet#queries
        fireEvent.click(getAllByRole('button')[0]);
        fireEvent.click(getAllByRole('button')[1]);
        fireEvent.click(getAllByRole('button')[2]);

        // https://testing-library.com/docs/dom-testing-library/api-async
        // await act(async () => {
        await waitForElement(() => baseElement.querySelector(`.${transitionFade}-entry-done`));
        // });

        expect(baseElement).toMatchSnapshot();
    });
});
