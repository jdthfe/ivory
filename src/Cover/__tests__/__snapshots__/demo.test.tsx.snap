// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoverDemo Cover change visible 1`] = `
<body>
  <div>
    <div
      class="eui-wingblank Cover eui-wingblank-m"
    >
      <h4>
        Cover
      </h4>
      <button
        class="eui-button eui-button-l eui-button-customize eui-button-ghost"
        role="button"
      >
        Cover
      </button>
      <div
        class="eui-cover"
      />
      <div
        class="eui-whitespace eui-whitespace-m"
      />
      <button
        class="eui-button eui-button-l eui-button-customize eui-button-ghost"
        role="button"
      >
        Cover | transparent
      </button>
      <div
        class="eui-cover eui-cover-transparent"
      />
      <div
        class="eui-whitespace eui-whitespace-m"
      />
      <h4>
        Cover.Transition
      </h4>
      <button
        class="eui-button eui-button-l eui-button-customize eui-button-ghost"
        role="button"
      >
        Cover.Transition | Portal
      </button>
    </div>
  </div>
  <div
    class="eui-cover eui-fade-entry-done eui-fade"
  />
</body>
`;

exports[`renders Cover correctly 1`] = `
<div>
  <div
    class="eui-wingblank Cover eui-wingblank-m"
  >
    <h4>
      Cover
    </h4>
    <button
      class="eui-button eui-button-l eui-button-customize eui-button-ghost"
      role="button"
    >
      Cover
    </button>
    <div
      class="eui-whitespace eui-whitespace-m"
    />
    <button
      class="eui-button eui-button-l eui-button-customize eui-button-ghost"
      role="button"
    >
      Cover | transparent
    </button>
    <div
      class="eui-whitespace eui-whitespace-m"
    />
    <h4>
      Cover.Transition
    </h4>
    <button
      class="eui-button eui-button-l eui-button-customize eui-button-ghost"
      role="button"
    >
      Cover.Transition | Portal
    </button>
  </div>
</div>
`;
